armory/art_experimental/attacks/snr_pgd.py

```diff
@@ -0,0 +1,147 @@
+            return x
+
+        eps = x_l2 * self.snr_sqrt_reciprocal
+        eps_step = eps * self.step_fraction
+        self.set_params(eps=eps, eps_step=eps_step)
+        return super().generate(x, y=y, **kwargs)
+
+    def _compute_perturbation(self, x, y, mask):
+        """
+        Compute perturbations.
+        :param x: Current adversarial examples.
+        :param y: Target values (class labels) one-hot-encoded of shape `(nb_samples, nb_classes)` or indices of shape
+                  (nb_samples,). Only provide this parameter if you'd like to use true labels when crafting adversarial
+                  samples. Otherwise, model predictions are used as labels to avoid the "label leaking" effect
+                  (explained in this paper: https://arxiv.org/abs/1611.01236). Default is `None`.
+        :param mask: An array with a mask to be applied to the adversarial perturbations. Shape needs to be
+                     broadcastable to the shape of x. Any features for which the mask is zero will not be adversarially
+                     perturbed.
+        :return: Perturbations.
+        """
+        import torch  # lgtm [py/repeated-import]
+
+        # Get gradient wrt loss; invert it if attack is targeted
+        import art
+
+        if art.__version__.startswith("1.4"):
+            grad = self.estimator.loss_gradient_framework(x, y) * (
+                1 - 2 * int(self.targeted)
+            )
+        else:
+            grad = self.estimator.loss_gradient(x=x, y=y) * (1 - 2 * int(self.targeted))
+        assert x.shape == grad.shape
+
+        # Apply mask
+        if mask is not None:
+            grad = grad * mask
+
+        # Remove NaNs
+        grad[torch.isnan(grad)] = 0.0
+        return grad
+
+    def _apply_perturbation(self, x, grad, eps_step):
+        """
+        Apply perturbation on examples.
+        :param x: Current adversarial examples.
+        :param grad: Current gradient.
+        :param eps_step: Attack step size (input variation) at each iteration.
+        :return: Adversarial examples.
+        """
+        import torch  # lgtm [py/repeated-import]
+
+        # Pick a small scalar to avoid division by 0
+        tol = 10e-8
+
+        # Apply norm bound
+        ind = tuple(range(1, len(x.shape)))
+
+        normalization = (
+            torch.sqrt(torch.sum(grad * grad, axis=ind, keepdims=True)) + tol
+        )
+        # eps_step > 0, normalization > 0; either could be inf
+        if eps_step == float("inf"):  # ignore normalization
+            perturbation = grad * eps_step
+        else:
+            if not torch.isfinite(normalization).all():
+                logger.warning(
+                    "Some gradient values are infinite. Perturbation will be 0!"
+                )
+            perturbation = grad * eps_step / normalization
+        perturbation[torch.isnan(perturbation)] = 0.0
+
+        x = x + perturbation
+        x[torch.isnan(perturbation)] = 0.0
```

Comment:
This should be extraneous due to zeroing out NaNs in `perturbation` a few lines earlier.